thorlcr/slave/traceslave.hpp

```diff
@@ -0,0 +1,219 @@
+        {
+            rowBatchCount = seqOfLastRow/rowBufferSize;
+        }
+        inline unsigned getSequence(void) const
+        {
+            return (rowBatchCount * rowBufferSize) + rowTailp;
+        }
+        inline void enqueue(const void *row, __int64 _cycleStamp)
+        {
+            // NOTE - updates then sets rowHeadp (atomically) so that the code
+            // below (ignoring oldest record) is safe
+            rowBuffer[rowTailp].set(row, _cycleStamp);
+            unsigned nextTailp = rowTailp + 1;
+            if (nextTailp==rowBufferSize)
+            {
+                nextTailp = 0;
+                ++rowBatchCount;
+            }
+            rowTailp = nextTailp;
+        }
+        inline void enqueue(const CThorRowHistory &_rowBuffer)
+        {
+            enqueue(_rowBuffer.get(),_rowBuffer.getCycleStamp());
+        }
+        // Move items in current queue into a second queue (clear out 'this' queue)
+        // NOTE - set skipLast to true when the oldest item must be skipped.  This should
+        // be used where the queue COULD have been updated after the buffers are swapped and
+        // the enqueueRowForTrace overwriting the oldest entry.
+        void queueOut(TraceQueue &traceOutQueue, bool skipOldestEntry)
+        {
+            const unsigned numberRowsToSkip = skipOldestEntry ? 1 : 0;
+            unsigned rowPos = rowTailp + numberRowsToSkip;
+            if (rowPos==rowBufferSize) rowPos = 0;
+            for (unsigned n=numberRowsToSkip; n<rowBufferSize;++n)
+            {
+                if (rowBuffer[rowPos])
+                {
+                    traceOutQueue.enqueue(rowBuffer[rowPos]);
+                    rowBuffer[rowPos].clear();
+                }
+                if (++rowPos==rowBufferSize) rowPos = 0;
+            }
+            traceOutQueue.setBatchCountFromSequence(getSequence());
+        }
+        // Dump items to MemoryBuffer (without clearing items from queue)
+        void dump(MemoryBuffer &mb, IHThorArg *helper) const
+        {
+            CommonXmlWriter xmlwrite(XWFnoindent|XWFtrim);
+            IOutputMetaData *meta = helper->queryOutputMeta();
+            const __int64 cyclesNow = get_cycles_now();
+            const unsigned lastSequence = getSequence();
+            unsigned rowPos = rowTailp;
+            for (unsigned n=0; n<rowBufferSize;++n)
+            {
+                const CThorRowHistory &rowCurrent = rowBuffer[rowPos];
+                if (rowCurrent)
+                {
+                    xmlwrite.outputBeginNested("Row", true);
+                    const unsigned seq = lastSequence - (rowBufferSize - n);
+                    const __int64 age = cycle_to_millisec(cyclesNow-rowCurrent.getCycleStamp());
+                    xmlwrite.outputInt(seq, sizeof(seq), "@seq");
+                    xmlwrite.outputInt(meta->getRecordSize(rowCurrent.get()), sizeof(size32_t), "@size");
+                    xmlwrite.outputInt(age, sizeof(age), "@age");
```

Comment:
>Roxie/HThor doesn't output the timestamp of the row.

Ah ok, thought it did.

Timestamp with milliscond definition is more useful, so can tell when a row was output and how long between outputted rows.
There other places we keep duration only and want to convert over to timestamps in fact